### PR TITLE
Add action manager processor

### PR DESCRIPTION
**Pull request by neithernut**
This PR introduces an initial implementation of the processor, which is the core of the aciton manager.
Targets #235.

### CODE REVIEW

File: src/action/processor.c [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+/*
+ *
+ * Forward declarations
+ *
+ */
+
+/**
+ * Prepare arguments for command execution
+ *
+ * This function prepares the stack for command ecxecution by copying everything
```
Comment (matthiasbeyer) [inline src/action/processor.c]: Typo.

File: src/action/processor.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+ * to the right position.
+ */
+static int
+ws_processor_prepare_args(
```
Comment (matthiasbeyer) [inline src/action/processor.c:48]: As this is a file-local function, please remove the `ws_` prefix.

File: src/action/processor.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+ *
+ */
+
+int
+ws_processor_init(
+    struct ws_processor* self,
+    struct ws_processor_stack* stack,
+    struct ws_transaction_command_list* commands
+) {
+    // initialize all the fields
```
Comment (matthiasbeyer) [inline src/action/processor.c:82]: init

**Comment by matthiasbeyer** [comment]
I didn't do a _that_ careful review, but the things look good. That's _NOT_ an ACK. I will do a review tomorrow without beeing tired then!

File: src/action/processor.c (hunk 3 above)
Comment (xVin) [inline src/action/processor.c:82]: lol :+1:

File: src/action/processor_stack.h [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+// forward declarations
+struct ws_value;
+union ws_value_union;
+
+/**
+ * A processor's stack
+ */
+struct ws_processor_stack {
+    union ws_value_union* data; //!< @private basepointer of the stack
```
Comment (matthiasbeyer) [inline src/action/processor_stack.h:47]: Why does this need to be a `ws_value_union`? You have the base type, `ws_value`, which you can use as ptr here!

File: src/values/union.c [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    case WS_VALUE_TYPE_NIL:
+        ws_value_nil_init(&dest->nil);
+        return 0;
+
+    case WS_VALUE_TYPE_BOOL:
+        ws_value_bool_init(&dest->bool_);
+        {
+            bool buf = ws_value_bool_plain(&dest->bool_);
+        return ws_value_bool_set(&dest->bool_, buf);
```
Comment (matthiasbeyer) [inline src/values/union.c]: Indentation.

File: src/values/union.c [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            return ws_value_int_set(&dest->int_, buf);
+        }
+
+    case WS_VALUE_TYPE_STRING:
+        ws_value_string_init(&dest->string);
+        {
+            struct ws_string* buf = ws_value_string_get_str(&dest->string);
+            ws_value_string_set_str(&dest->string, buf);
+            ws_object_unref(&buf->obj);
+        }
```
Comment (matthiasbeyer) [inline src/values/union.c:74]: Here is a return missing, right?

File: src/action/processor_stack.h (hunk 4 above)
Comment (neithernut) [inline src/action/processor_stack.h:47]: I think I made mayself clear about this when I _introduced_ the `ws_value_union` type:
Nope, I'll not add malloc calls for each nil and int value we need!

**Comment by neithernut** [comment]
Before this eventually gets merged, I'd like to bring in a proposal, which may result in some "extra" work within this PR:

The overal idea is that functions receive their argumetns as values being pushed on the stack. Currently, you may either give arguments directly or by a position on the stack as it was before the command involcation. This PR introduces a TODO, which suggests that this may be optimized in a way that allows calling functions without moving around values to construct the stack, however, it's not possible to do this without introducing some clusterfuck.

However, if we allow a statement to specify that it will take the topmost `n` values of the stack as arguments, things are suddenly very nice.

TL;DR: We should have statements which specify the number of arguments without shipping the actual argument list.

**Comment by matthiasbeyer** [comment]
What kind of benefits would result from a change like that, when talking of user interaction with the API?

**Comment by neithernut** [comment]
Evaluation of expressions is a whole lot faster, and more intuitive for people who know Reverse Polish notation. It also eases things considerably if we come up with a DSL.

**Comment by matthiasbeyer** [comment]
I assume that the request/transaction steps would look like this:
- I want 3 arguments from the stack for the next command
- The next command

right? If yes, this sounds sane to me. But it would also cause changes in the JSON parser, wouldn't it? If yes, I would recommend add such functionality later. (Execution/Parsing) Speed is not our only concern at this moment!

**Comment by neithernut** [comment]
It's also a question of a sane interface!

I submitted a PR (#238). As you'll see, the changes in the interface are purely in semantics and can therefore be introduced without any problems emerging (as long as I perform the neccessary changes _first_, which I intent).

**Comment by neithernut** [comment]
I rebased and added the new feature. in `ws_processor_exec()`

File: src/action/processor.c [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        struct ws_value* src = NULL;
+        switch (cur_arg->type) {
+        case direct:
+            src = cur_arg->arg.val;
+
+        case indirect:
+            src = ws_processor_stack_value_at(stack, cur_arg->arg.pos);
+        }
+
+        if (!src) return -EINVAL;
```
Comment (matthiasbeyer) [inline src/action/processor.c]: Missing block around return statement here.

File: src/action/processor.c [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+size_t
+ws_processor_jump(
+    struct ws_processor* self,
+    size_t value
+) {
+    //calculate the current position
+    size_t target_pos = (self->pc - self->commands->statements) + value;
+
+    // perform bound check
+    if (target_pos >self->commands->n) {
```
Comment (matthiasbeyer) [inline src/action/processor.c]: Missing space after `>`

File: src/action/processor.h [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+;
+
+/**
+ * Perform a jump (by incrementing the program counter)
+ *
+ * @return 0 if the jump was successful, a positive value if the jump would
+ *           result in a position outside the current statement list.
+ */
+size_t
+ws_processor_jump(
```
Comment (matthiasbeyer) [inline src/action/processor.h:103]: Why needs this to be exported?

File: src/action/processor.h [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+ws_processor_exec(
+    struct ws_processor* self //!< command processor context
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Perform a jump (by incrementing the program counter)
+ *
+ * @return 0 if the jump was successful, a positive value if the jump would
```
Comment (matthiasbeyer) [inline src/action/processor.h:99]: `"would"` ? So it doesn't actually do something if the jump would result out of the statement list? Please clearify!

File: src/action/processor_stack.h [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    size_t size; //!< @private size of the stack
+    size_t top; //!< @private top of the stack, as position from the basepointer
+    size_t frame; //!< @private base pointer of the current frame
+};
```
Comment (matthiasbeyer) [inline src/action/processor_stack.h:51]: As all fields in the processor stack type: Why not hiding this type implementation?

File: src/action/processor_stack.h [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    size_t slots //!< number of slots to pop
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Get the top of the stack
+ *
+ * This function returns the top of the stack, which actually is an invalid
+ * value.
```
Comment (matthiasbeyer) [inline src/action/processor_stack.h:112]: Why would you want an invalid value from the stack? Maybe add some more information here why it can be useful.

File: src/action/processor_stack.h [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+ * If you subtract from that pointer, you get valid values _on_ the stack.
+ * At some point, you reach the bottom of the stack.
```
Comment (matthiasbeyer) [inline src/action/processor_stack.h:115]: How to identify the bottom of the stack when doing this?

File: src/action/processor_stack.h [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    struct ws_processor_stack* self //!< the stack
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Get the bottom of the stack
+ *
+ * This function returns the bottom of the stack.
+ * If you add to that pointer, you move toward the top of the stack.
```
Comment (matthiasbeyer) [inline src/action/processor_stack.h:130]: How to identify that we are on the last valid value?

File: src/action/processor_stack.h [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+
+/**
+ * Get a value at a given position
+ *
+ * @return value at given position
+ */
+struct ws_value*
+ws_processor_stack_value_at(
+    struct ws_processor_stack* self, //!< the stack
+    ssize_t pos //!< position of value
```
Comment (matthiasbeyer) [inline src/action/processor_stack.h:149]: Why can this be negative? Please add a short note! (I understand why, but well...)

File: src/action/processor_stack.h [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+ws_processor_stack_start_frame(
+    struct ws_processor_stack* self //!< the stack
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Restore a frame
+ *
+ * Call this function to restore a frame previously started using
```
Comment (matthiasbeyer) [inline src/action/processor_stack.h:169]: What does "restore" mean in this context?

File: src/action/processor_stack.c [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+#include <malloc.h>
+#include <string.h>
+
+#include "action/processor_stack.h"
+#include "values/nil.h"
+#include "values/union.h"
+#include "values/value.h"
+
+
+#define INITIAL_STACK_SIZE (4)
```
Comment (matthiasbeyer) [inline src/action/processor_stack.c:39]: Is this actually a sane value? I consider it as much to less, but well, we can optimize that later on very easily, I guess.

Comment (matthiasbeyer) [inline src/action/processor_stack.c:39]: Ah, just saw you do a `*2` reallocation on resize. Nuff said.

File: src/action/processor.h (hunk 9 above)
Comment (neithernut) [inline src/action/processor.h:103]: You use this function in the "jump" command.

File: src/action/processor_stack.c [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+void
+ws_processor_stack_pop(
+    struct ws_processor_stack* self,
+    size_t slots
+) {
+    // check whether there are at least that many values on the stack
+    if (slots > (self->top - self->frame)) {
+        slots = self->top - self->frame;
```
Comment (matthiasbeyer) [inline src/action/processor_stack.c]: If I understand things right: If `slots` is bigger than the number of values on the stack, we only pop the maximum value. This should be stated in the documentation of the function, I guess.

File: src/action/processor_stack.c [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    size_t new_top  = self->top - slots;
+    while ((new_top > new_size) && (new_top > INITIAL_STACK_SIZE)) {
+        new_size /= 2;
+    }
+
+    // deinitialize all the values on the way down
+    size_t cur = self->top;
+    while (cur > new_top) {
+        --cur;
+        ws_value_deinit(&self->data[cur].value);
```
Comment (matthiasbeyer) [inline src/action/processor_stack.c:132]: You deinitialize here, but shouldn't be the values be initialized to `nil` afterwards? That's what you do when growing the stack, so I assume you should do it here, too, right?

File: src/action/processor_stack.c [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    self->top = new_top;
+
+    // check whether we may resize
+    if (new_size != self->size) {
+        // try to reallocate
+        union ws_value_union* new_data;
+        new_data = realloc(self->data, sizeof(*(self->data)) * new_size);
+        if (!new_data) {
+            // noncritical error
+            return;
```
Comment (matthiasbeyer) [inline src/action/processor_stack.c]: Would be shorter if the if wouldn't use an early return and you would set the members if the reallocation succeeded. _duckandrun_

File: src/action/processor_stack.c [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+
+    return self->data;
+}
+
+struct ws_value*
+ws_processor_stack_value_at(
+    struct ws_processor_stack* self,
+    ssize_t pos
+) {
+    size_t top = self->top;
```
Comment (matthiasbeyer) [inline src/action/processor_stack.c]: I would embedd the `self->top` into the code below, it's not that much overhead to read.

File: src/values/union.h [hunk 22]
```diff
@@ -25,6 +25,16 @@
  * along with waysome. If not, see <http://www.gnu.org/licenses/>.
  */
 
+
+/**
+ * @addtogroup values "Value types"
+ *
+ * @{
+ */
+
```
Comment (matthiasbeyer) [inline src/values/union.h:34]: Maybe we should add another documentation group within the values group for internal stuff. (I consider that the union value type is not exported through the API, right?

File: src/values/union.h [hunk 23]
```diff
@@ -37,8 +47,21 @@
...
-    struct ws_value_bool    bool;   //!< bool value
-    struct ws_value_int     int;    //!< int value
+    struct ws_value_bool    bool_;  //!< bool value
+    struct ws_value_int     int_;   //!< int value
     struct ws_value_string  string; //!< string value
 };
 
+/**
+ * Initializes the union by copying another value
+ */
```
Comment (matthiasbeyer) [inline src/values/union.h:61]: Missing: Return documentation, documentation why this does nothing if the value type is either "value" or a "non-value" type.

**Comment by matthiasbeyer** [comment]
Finally I managed to do a complete review on this. The code looks really sane for me and I like what you introduced. So please, don't feel bad about my comments here!

File: src/action/processor_stack.c (hunk 20 above)
Comment (neithernut) [inline src/action/processor_stack.c]: No, it wouldn't.

Comment (matthiasbeyer) [inline src/action/processor_stack.c]: It would result in 3 less lines in the function.

File: src/action/processor_stack.h (hunk 11 above)
Comment (neithernut) [inline src/action/processor_stack.h:51]: Because it will be allocated on the stack, not on the heap.

File: src/action/processor_stack.h (hunk 16 above)
Comment (neithernut) [inline src/action/processor_stack.h:169]: Seriously? Should be pretty obvious.

File: src/action/processor_stack.c (hunk 19 above)
Comment (neithernut) [inline src/action/processor_stack.c:132]: Nope, values which are not on the stack _must_ be invalid, so we can determine the number of arguments a function gets by reading up to the first invalid argument.

File: src/values/union.h (hunk 22 above)
Comment (neithernut) [inline src/values/union.h:34]: Of course it's not. But managing the documentation groups is nothing that should go in this PR.

File: src/action/processor_stack.h (hunk 14 above)
Comment (neithernut) [inline src/action/processor_stack.h:130]: `WS_VALUE_TYPE_NONE`

File: src/action/processor_stack.h (hunk 15 above)
Comment (neithernut) [inline src/action/processor_stack.h:149]: **not on this scope**
This belongs in an more extensive external documentation.

File: src/action/processor_stack.h (hunk 13 above)
Comment (neithernut) [inline src/action/processor_stack.h:115]: Coparision to the pointer you get from `ws_processor_stack_bottom()`

File: src/action/processor_stack.h (hunk 15 above)
Comment (matthiasbeyer) [inline src/action/processor_stack.h:149]: So add an `@see` to this documentation.

File: src/action/processor_stack.h (hunk 14 above)
Comment (matthiasbeyer) [inline src/action/processor_stack.h:130]: Should be in the doc.

File: src/action/processor_stack.h (hunk 15 above)
Comment (neithernut) [inline src/action/processor_stack.h:149]: I have to write that documentation **first**.

**Comment by neithernut** [comment]
I added the first chunk of documentation of stack semantics. More to come.

File: src/action/processor_stack.h (hunk 15 above)
Comment (matthiasbeyer) [inline src/action/processor_stack.h:149]: Of course.

File: src/action/stack.md [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+ |        | <- top
+ | FOO    |
+ | BAR    | <- argument of the function implementation
+ | BAZ    |
+ | DUCK   | <- bottom
+ +--------+
+
+After a function invocation, all the arguments are popped from the stack except
+the first one, which is the return value of the function.
+
```
Comment (matthiasbeyer) [inline src/action/stack.md:84]: This belongs into `doc/` in my opinion. I will open an Issue for this to discuss.

**Comment by neithernut** [comment]
I outsources a few lines of code from `ws_processor_exec()` to another function and fixed and expanded the documentation a little.
Please review again and tell me if there are any questions left.

**Comment by neithernut** [comment]
Updated because of #243.

File: src/action/stack.md [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+top, `-1` being the topmost value, `-2` the next lower one and so on.
+The following graph shows a stack with four items and the position of each one
+of them in both direct and relative addressing:
+
+    +------+
+    |      |         <- top
+    | FOO  | -1 or 3
+    | BAR  | -2 or 2
+    | BAZ  | -3 or 1
+    | DUCK | -4 or 0 <- bottom
```
Comment (matthiasbeyer) [inline src/action/stack.md:22]: :+1: for "duck"

File: src/action/processor_stack.h (hunk 16 above)
Comment (matthiasbeyer) [inline src/action/processor_stack.h:169]: It is now, after I read your documentation. Please add a `@see` in the function documentation, refering to the documentation document.

Comment (neithernut) [inline src/action/processor_stack.h:169]: The type documentation already has a `@see` tag.

**Comment by matthiasbeyer** [comment]
ACK.

For the documentation thing: I will add a note in the related issue.
